### ISSUE TITLE
TypeError error while using real-time parsing

### ISSUE THREAD

**Issue by RedFantom** [RedFantom/GSF-Parser#20]
In line 113 of `realtime.py` a TypeError occurs in the function `parse(...)` at some point while using real-time parsing. Particularly, the program is using the `__getitem__` function of the `line` parameter, a returned object from the function `line_to_dictionary(line)` in the same file. The full error is

`’NoneType’ object has no attribute ’groups’`
`Exception in thread Thread-1:`
`Traceback (most recent call last):`
`File ”threading.py”, line 801, in __bootstrap_inner`
`File ”stalking_alt.py”, line 56, in run`
`File ”realtimeframe.py”, line 198, in callback`
`File ”realtime.py”, line 113, in parse`
`TypeError: ’NoneType’ object has no attribute ’__getitem__’`

There is a `[DEBUG]` message for this, particularly in line 104 of the `realtime.py` file, which was used during debugging. However, clearly the `line_to_dictionary(line)` function still returns a `None` in some cases. Further investigation is required.
After the error, the real-time parsing process stops and the statistics do not update anymore as the Thread has encountered and error and stops.

**Comment by naiii** [comment]
Ok. If you have a bit of patience I will take a look. Thursday I have an exam, but after that, I might have a bit of time to look at it (the next and last one is a week later).

**Comment by RedFantom** [comment]
Thank you, that would be much appreciated! :) Good luck with your exam!

**Comment by naiii** [comment]
Can you describe in which circumstances this error occurs? (If I were to replicate this error, what should I do?)

**Comment by RedFantom** [comment]
While using real-time parsing myself yesterday, I probably also got the same error (the real-time parsing stopped but I did not have a debug window). This only happened while playing SWTOR:GSF, it did not occur during about half an hour of using the `LogGenerator`, and earlier tests that lasted even longer also with the `LogGenerator`. I'm afraid the only way to replicate the error is by playing GSF. If you don't have time to do so yourself, I go flying again tomorrow.

**Comment by naiii** [comment]
That would be cool. Maybe add a debug line in the `except` clause saying you are there, and what the line (function argument) is. This way I know by what kind of line the error occurs.

**Comment by naiii** [comment]
Ok. I added a debug print for this issue, where I think the problem is happening.
Could you please try it now, and say (if when it crash this line is printed) what it outputs (copy paste the line) or if it actually didn't print anything new. Thanks!

**Comment by RedFantom** [comment]
Caught it! The error occurred when leaving the match and going back to the Imperial Fleet. This was the full error print:
`Exception in thread Thread-1:`
`Traceback (most recent call last):`
`  File "C:\Python_2.7\lib\threading.py", line 801, in __bootstrap_inner`
`    self.run()`
`  File "C:\Users\user\Source Code\GSF-Parser\stalking_alt.py", line 56, in run`
`    self.callback(self.read_from_file())`
`  File "C:\Users\user\Source Code\GSF-Parser\realtimeframe.py", line 198, in callback`
`    self.parser.parse(process)`
`  File "C:\Users\user\Source Code\GSF-Parser\realtime.py", line 113, in parse`
`    if(line['source'] == line['destination'] and "@" not in line['source'] and ":" not in line['source'] and`
`TypeError: 'NoneType' object has no attribute '__getitem__'`
And then the part you requested:
`('[DEBUG] line_to_dictionary(): arg:', '[19:04:08.267] [49250', "'tuple = group.groups()' error raised, with` `group: ", None)`
`'NoneType' object has no attribute 'groups'`
`[DEBUG] Line is of NoneType`
And the CombatLog in question is attached: [Snippet.txt](https://github.com/RedFantom/GSF-Parser/files/737708/Snippet.txt)

**Comment by naiii** [comment]
Ok. it appears that  the error does not occur in the `line_to_directory()` function but at line 113, which then causes `line_to_dictionary()` to crash as it does not get a full line to pars.

I see that you added a couple of line to the parse() function (line 113) included. I have a couple of question:

- Why do you do this checks?
`if ... and ":" not in line['source'] and not bool(re.search(r'\d', line['source'])):` (line 113)

- More a curiosity question, what does this do?
`self.insert(line, vars.rt_timing, self.active_id)`

**Comment by RedFantom** [comment]
All right, so it does not get a full line to parse. That might mean that the parser actually reads the line from SWTOR while the IO buffer is busy flushing to the file. Having the `parse()` function wait for the rest of the line is not an option, as it will never receive it. This problem is in the `stalking_alt.py` file, specifically in `read_from_file(self)` in line 66. Honestly, I don't know how to fix this than to work around it by passing if the line is of `NoneType`...

To answer your question: `self.insert(line, vars.rt_timing, self.active_id)` calls the function (callback) that first runs the line through the `statistics.pretty_event(line_dict, start_of_match, active_id)` function in order to take the resulting string (and colour) and put it in the events listbox of real-time parsing. `line` is a dictionary of a line from `line_to_dictionary()`, `vars.rt_timing` is a datetime that contains the start of the match and `active_id` is a string containing the current ID number of the player.

**Comment by naiii** [comment]
Ok. Thanks.

I did run the snippet you privided with the log simulator and it did crash after the line (this is the last line parsed before the crash):
[19:04:50.347] [@Redfantom] [@Redfantom] [Sprint {810670782152704}] [ApplyEffect {836045448945477}: Sprint {810670782152704}] ()

This is the trace:```

Traceback (most recent call last):
  File "/home/century/Documents/Programming/GSF-Generator/Parser.py", line 549, in <module>
    stalker.loop()
  File "/home/century/Documents/Programming/GSF-Generator/LogStalking.py", line 89, in loop
    self.readlines(file)
  File "/home/century/Documents/Programming/GSF-Generator/LogStalking.py", line 196, in readlines
    self._callback(file.name, lines)
  File "/home/century/Documents/Programming/GSF-Generator/Parser.py", line 478, in callback
    process = line_to_dictionary(line)
  File "/home/century/Documents/Programming/GSF-Generator/Parser.py", line 499, in line_to_dictionary
    log['ability'] = log['ability'].rsplit(None, 1)[1][1:-1]
IndexError: list index out of range
```

I will investigate further as soon as I can.

**Comment by RedFantom** [comment]
The commit I authored yesterday (4d61bb0 and 5c5e8d1 to fix a problem with that commit) makes sure that each line passed to the `Parser` object by the `LogStalker` object is a line that actually passes the `line_to_dictionary(...)` function in `realtime.py` by taking the last line and checking whether it passes. The problem exists because the game is sometimes still writing and then the line is not fully read, making for a `NoneType` returned from the `line_to_dictionary(...)` function. Only if the last line returns something else than `None` from this function, it is added to the list of lines returned and only then is it also added to the total of `self.read_so_far` in order to read it again if the line was not fully written by the game yet. I tested yesterday, and it seems to fix it, but I will run further tests.
